# Hacker News Analysis


We'll compare these two types of posts to determine the following:

Do Ask HN or Show HN receive more comments on average?
Do posts created at a certain time receive more comments on average?

In [4]:
from csv import reader
opened_file=open('hacker_news.csv')
read_file=reader(opened_file)
hn=list(read_file)
headers=hn[0]
hn=hn[1:]
print(hn[:5])
print(headers)

[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]
['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


In [7]:
ask_posts=[]
show_posts=[]
other_posts=[]
for row in hn:
    title=row[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
print('posts in ask posts: ' , len(ask_posts))
print('posts in show posts: ' , len(show_posts))
print('posts in other posts: ' , len(other_posts))

posts in ask posts:  1744
posts in show posts:  1162
posts in other posts:  17194


In [9]:
total_ask_comments=0
for element in ask_posts:
    total_ask_comments+=int(element[4])
avg_ask_comments=total_ask_comments/len(ask_posts)
print(avg_ask_comments)
    

14.038417431192661


In [10]:
total_show_comments=0
for element in show_posts:
    total_show_comments+=int(element[4])
avg_show_comments=total_show_comments/len(show_posts)
print(avg_show_comments)

10.31669535283993


ASK Post receive more comments on an average as compared to SHOW Post comments
Average comments on ask post : 14.038417431192661
Average comments on show post : 10.31669535283993

In [11]:
import datetime as dt

In [16]:
result_list=[]
for row in ask_posts:
    inner_list=[]
    inner_list.append(row[6])
    inner_list.append(int(row[4]))
    result_list.append(inner_list)
result_list[1]

['11/22/2015 13:43', 29]

In [17]:
counts_by_hour={}
comments_by_hour={}
for row in result_list:
    date_obj=dt.datetime.strptime(row[0],"%m/%d/%Y %H:%M")
    hour=date_obj.strftime("%H")
    if hour in counts_by_hour:
        counts_by_hour[hour]+=1
        comments_by_hour[hour]+=row[1]
    else:
        counts_by_hour[hour]=1
        comments_by_hour[hour]=row[1]

In [20]:
counts_by_hour


{'09': 45,
 '13': 85,
 '10': 59,
 '14': 107,
 '16': 108,
 '23': 68,
 '12': 73,
 '17': 100,
 '15': 116,
 '21': 109,
 '20': 80,
 '02': 58,
 '18': 109,
 '03': 54,
 '05': 46,
 '19': 110,
 '01': 60,
 '22': 71,
 '08': 48,
 '04': 47,
 '00': 55,
 '06': 44,
 '07': 34,
 '11': 58}

In [21]:
comments_by_hour

{'09': 251,
 '13': 1253,
 '10': 793,
 '14': 1416,
 '16': 1814,
 '23': 543,
 '12': 687,
 '17': 1146,
 '15': 4477,
 '21': 1745,
 '20': 1722,
 '02': 1381,
 '18': 1439,
 '03': 421,
 '05': 464,
 '19': 1188,
 '01': 683,
 '22': 479,
 '08': 492,
 '04': 337,
 '00': 447,
 '06': 397,
 '07': 267,
 '11': 641}

In [22]:
avg_by_hour=[]
for key,value in counts_by_hour.items():
    inner_list=[]
    inner_list.append(key)
    avg=comments_by_hour[key]/value
    inner_list.append(avg)
    avg_by_hour.append(inner_list)
avg_by_hour

[['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['05', 10.08695652173913],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['00', 8.127272727272727],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034]]

In [24]:
swap_avg_by_hour=[]
for row in avg_by_hour:
    inner_list=[]
    inner_list.append(row[1])
    inner_list.append(row[0])
    swap_avg_by_hour.append(inner_list)
swap_avg_by_hour

[[5.5777777777777775, '09'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [16.796296296296298, '16'],
 [7.985294117647059, '23'],
 [9.41095890410959, '12'],
 [11.46, '17'],
 [38.5948275862069, '15'],
 [16.009174311926607, '21'],
 [21.525, '20'],
 [23.810344827586206, '02'],
 [13.20183486238532, '18'],
 [7.796296296296297, '03'],
 [10.08695652173913, '05'],
 [10.8, '19'],
 [11.383333333333333, '01'],
 [6.746478873239437, '22'],
 [10.25, '08'],
 [7.170212765957447, '04'],
 [8.127272727272727, '00'],
 [9.022727272727273, '06'],
 [7.852941176470588, '07'],
 [11.051724137931034, '11']]

Top 5 Hours for Ask Posts Comments

In [25]:
sorted_swap=sorted(swap_avg_by_hour,reverse=True)
print(sorted_swap[:5])

[[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21']]


In [27]:
for element in sorted_swap[:5]:
    time1=dt.datetime.strptime(element[1],"%H")
    time2=time1.strftime("%H:%M")
    template="{}: {cnt:.2f} average comments per post"
    template=template.format(time2,cnt=element[0])
    print(template)

15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post
